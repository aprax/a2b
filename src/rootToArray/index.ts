export interface TreeNode {
  val: Object;
  left?: TreeNode | undefined;
  right?: TreeNode | undefined;
}

const rootToArray = (
  result: string[],
  node: TreeNode,
  parentIndex?: number,
  isLeft = false
) => {
  let index = 0;
  if (typeof parentIndex === "number") {
    index = isLeft ? 2 * parentIndex + 1 : 2 * parentIndex + 2;
  }

  result[index] = node.val.toString();
  if (node.left) {
    rootToArray(result, node.left, index, true);
  }
  if (node.right) {
    rootToArray(result, node.right, index, false);
  }
};
export default rootToArray;
