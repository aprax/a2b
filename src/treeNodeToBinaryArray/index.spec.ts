import rootToArray, { TreeNode } from ".";

it("should match [1,2]", () => {
  const root: TreeNode = {
    val: 1,
    left: { val: 2 },
  };

  const actual = rootToArray(root);
  expect(actual).toMatchObject(["1", "2"]);
});

it("should match [1,2,3]", () => {
  const root: TreeNode = {
    val: 1,
    left: { val: 2 },
    right: { val: 3 },
  };
  const actual = rootToArray(root);
  expect(actual).toMatchObject(["1", "2", "3"]);
});

it("should match [1,undefined,3]", () => {
  const root: TreeNode = {
    val: 1,
    right: { val: 3 },
  };
  const actual = rootToArray(root);
  expect(actual).toMatchObject(["1", undefined, "3"]);
});

(() => {
  const expected = [
    "1",
    "2",
    undefined,
    undefined,
    "5",
    undefined,
    undefined,
    undefined,
    undefined,
    undefined,
    "11",
  ];
  it(`should match ${expected}`, () => {
    const root: TreeNode = {
      val: 1,
      left: {
        val: 2,
        right: {
          val: 5,
          right: {
            val: 11,
          },
        },
      },
    };
    const actual = rootToArray(root);
    expect(actual).toMatchObject(expected);
  });
})();
