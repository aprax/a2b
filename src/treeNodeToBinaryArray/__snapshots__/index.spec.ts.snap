// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot for {"val":1,"left":{"val":2,"right":{"val":5,"right":{"val":11}}}} 1`] = `
Array [
  "1",
  "2",
  ,
  ,
  "5",
  ,
  ,
  ,
  ,
  ,
  "11",
]
`;

exports[`should match snapshot for {"val":3,"left":null,"right":{"val":9,"left":null,"right":null}} 1`] = `
Array [
  "3",
  ,
  "9",
]
`;
